lib/bolt/transport/ssh/connection.rb

```diff
@@ -148,119 +142,86 @@ def disconnect
           end
         end
 
-        def handled_sudo(channel, data, stdin)
-          if data.lines.include?(Sudoable.sudo_prompt)
-            if @sudo_password
-              channel.send_data("#{@sudo_password}\n")
-              channel.wait
-              return true
-            else
-              # Cancel the sudo prompt to prevent later commands getting stuck
-              channel.close
-              raise Bolt::Node::EscalateError.new(
-                "Sudo password for user #{@user} was not provided for #{target.safe_name}",
-                'NO_PASSWORD'
-              )
-            end
-          elsif data =~ /^#{@sudo_id}/
-            if stdin
-              channel.send_data(stdin)
-              channel.eof!
-            end
-            return true
-          elsif data =~ /^#{@user} is not in the sudoers file\./
-            @logger.debug { data }
-            raise Bolt::Node::EscalateError.new(
-              "User #{@user} does not have sudo permission on #{target.safe_name}",
-              'SUDO_DENIED'
-            )
-          elsif data =~ /^Sorry, try again\./
-            @logger.debug { data }
-            raise Bolt::Node::EscalateError.new(
-              "Sudo password for user #{@user} not recognized on #{target.safe_name}",
-              'BAD_PASSWORD'
-            )
-          end
-          false
-        end
-
-        def execute(command, sudoable: false, **options)
-          result_output = Bolt::Node::Output.new
-          run_as = options[:run_as] || self.run_as
-          escalate = sudoable && run_as && @user != run_as
-          use_sudo = escalate && @target.options['run-as-command'].nil?
+        def execute(command_str)
+          in_rd, in_wr = IO.pipe
+          out_rd, out_wr = IO.pipe
+          err_rd, err_wr = IO.pipe
+          th = Thread.new do
```

Comment:
I'm a bit confused about why we're opening a thread here...this is only executed on one target right? What else does threading get us?

Reply:
In order to have a single `sudo` implementation, we need both the local and SSH transports to present the same API. `Net::SSH` provides a blocking API that uses callbacks to send stdout and stderr, while the local transport uses `Open3.popen3` which returns `IO` objects for stdin, stdout and stderr, along with a `Process::Waiter` object that can be polled to wait for the command to finish and receive its exit code.

To allow the two transports to present the same API, we squish `Net::SSH` into the same paradigm as `Open3.popen3`. We create pipes for stdin, stdout and stderr and start a thread to internally wait for the command to finish. This thread effectively bridges between the pipe that `Bolt::Shell::Bash` is writing to and `Net::SSH` that doesn't use `IO` objects but just has a `send_data` method. Similarly, it bridges between the callbacks that `Net::SSH` uses for stdout/stderr and the pipes that the bash shell expects to read from.

Because the shell class and the connection class are both simultaneously reading from and writing to pipes to communicate with each other, they need to be running in separate threads.

Reply:
:scream: That makes sense but also sounds terrible